scalanet/src/io/iohk/scalanet/peergroup/dynamictls/ChannelAwareQueue.scala

```diff
@@ -0,0 +1,63 @@
+package io.iohk.scalanet.peergroup.dynamictls
+
+import io.iohk.scalanet.peergroup.CloseableQueue
+import io.netty.channel.ChannelConfig
+import monix.eval.Task
+import monix.execution.ChannelType
+
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+  * Wraps an underlying unbounded CloseableQueue queue and bounds it based on netty auto-read feature.
+  * While auto-read is disabled received messages start to accumulate in underlying os RCV tcp buffer. When RCV buffer is full
+  * sender SND buffer will start to buffer un-sent bytes. When sender SND buffer is full, the default behaviour is that
+  * write(xxx) will block.  In our case, sendMessage Task will not finish until there will be place in SND buffer
+  *
+  *
+  * WARNING: Actual limit may sometimes goes higher, as each netty read can return more than one message.
+  *
+  * @param limit how many items can accumulate in the queue
+  * @param queue is the underlying closeable message queue
+  */
+private[scalanet] final class ChannelAwareQueue[M] private (limit: Int, queue: CloseableQueue[M]) {
+  private val queueSize = new AtomicLong(0)
+
+  private val lowerBound: Int = Math.max(1, limit / 2)
+
+  def size: Long = queueSize.get()
+
+  def offer(channelConfig: ChannelConfig, a: M): Task[Unit] = {
```

Comment:
Why isn't the return type of `CloseableQueue.offer` preserved?

Reply:
because we never make us of it i.e in all call sites we were just voiding it. If we will ever need to expose full return type from `ChannelAwareQueue `for some reason it will be easy to bring it back.

Reply:
Yes but there's no reason for every queue abstraction to have an ever so slightly different interface. When it even says in the docs that this is a wrapper over that other type, they should be interchangeable.